### PR TITLE
Feature/vs 12109 : Adding Support for 3D Secure

### PR DESCRIPTION
**Pull request by Nicole8164**

### CODE REVIEW

File: src/Message/PaymentRequest.php [hunk 1]
```diff
@@ -203,14 +223,23 @@ protected function getPaymentParams($card)
...
             'paymentRequest.amount.value' => $this->getAmount(),
             'paymentRequest.reference' => $this->getTransactionReference(),
             'paymentRequest.shopperEmail' => $card->getEmail(),
-            'paymentRequest.shopperReference' => $card->getShopperReference(),
+            'paymentRequest.shopperReference' => $card->getShopperReference()
         ];
+
+        if ($this->get3dSecure()) {
+            $payment_params += [
+                'paymentRequest.browserInfo.userAgent' => $_SERVER['HTTP_USER_AGENT'],
```
Comment (Dandroas) [inline src/Message/PaymentRequest.php]: would it make sense to create a getter/setter for the globals in pass it in?

File: src/Message/SecureRequest.php [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+     *
+     * @return array
+     */
+    public function getData()
+    {
+        $returned_data = $this->httpRequest->request->all();
+        return [
+            "action" => "Payment.authorise3d",
+            "paymentRequest3d.merchantAccount" => $this->getMerchantAccount(),
+            "paymentRequest3d.browserInfo.userAgent" => $_SERVER['HTTP_USER_AGENT'],
```
Comment (Dandroas) [inline src/Message/SecureRequest.php]: would it make sense to create a getter/setter for the globals in pass it in?

File: tests/GatewayTest.php [hunk 3]
```diff
@@ -43,9 +48,9 @@ private function getPaymentParams()
                     'billing_state' => 'Ille dfrance',
                     'billing_country' => 'FR',
                     'email' => 'dandroas@gmail.com',
-                    'shopper_reference' => '123654'
+                    'shopper_reference' => '123654',
```
Comment (Dandroas) [inline tests/GatewayTest.php]: Why do we include the trailing commas?

Comment (Nicole8164) [inline tests/GatewayTest.php]: Not sure how that got added. It was a mistake will remove.

File: src/Message/PaymentResponse.php [hunk 4]
```diff
@@ -50,4 +60,37 @@ public function getCode()
...
+    {
+        return $this->data['paymentResult_issuerUrl'];
+    }
+
+    /**
+     * Returns the 3D Secure request data for the issuer
+     *
+     * @return string
+     */
+    public function getPaRequest()
```
Comment (mcinbell) [inline src/Message/PaymentResponse.php]: What's 'PA'?

Comment (Nicole8164) [inline src/Message/PaymentResponse.php]: I believe it's 'Payment authorisation' that's how it's described in the docs for the PaResponse. PaRequest just states `3D Secure request data for the issuer.`

**Pull request merged**